sdks/python/apache_beam/io/mongodbio_test.py

```diff
@@ -0,0 +1,171 @@
+    mock_tracker.assert_called_with(0, 30)
+    self.mongo_source.get_range_tracker(10, 20)
+    mock_tracker.assert_called_with(10, 20)
+
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_read(self, mock_client):
+    mock_tracker = mock.MagicMock()
+    mock_tracker.try_claim.return_value = True
+    mock_tracker.start_position.return_value = 0
+    mock_tracker.stop_position.return_value = 2
+
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value \
+      .__getitem__.return_value.find.return_value = [{'x': 1}, {'x': 2}]
+
+    result = []
+    for i in self.mongo_source.read(mock_tracker):
+      result.append(i)
+    self.assertListEqual([{'x': 1}, {'x': 2}], result)
+
+  def test_display_data(self):
+    data = self.mongo_source.display_data()
+    self.assertTrue('uri' in data)
+    self.assertTrue('database' in data)
+    self.assertTrue('collection' in data)
+
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test__get_avg_document_size(self, mock_client):
+    mock_client.return_value.__enter__.return_value.__getitem__ \
+      .return_value.command.return_value = {'avgObjSize': 5}
+    self.assertEqual(5, self.mongo_source._get_avg_document_size())
+
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_get_document_count(self, mock_client):
+    mock_client.return_value.__enter__.return_value.__getitem__ \
+      .return_value.__getitem__.return_value.count_documents.return_value = 10
+
+    self.assertEqual(10, self.mongo_source._get_document_count())
+
+
+class TestReadFromMongoDB(unittest.TestCase):
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_read_from_mongodb(self, mock_client):
+    objects = [{'x': 1}, {'x': 2}]
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      command.return_value = {'avgObjSize': 1}
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      __getitem__.return_value.find.return_value = objects
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      __getitem__.return_value.count_documents.return_value = 2
+
+    with TestPipeline() as p:
+      docs = p | 'ReadFromMongoDB' >> ReadFromMongoDB(
+          uri='mongodb://test', db='db', coll='collection')
+      assert_that(docs, equal_to(objects))
+
+
+class Test_GenerateObjectIdFn(unittest.TestCase):
+  def test_process(self):
+    with TestPipeline() as p:
+      output = (p | "Create" >> beam.Create([{
+          'x': 1
+      }, {
+          'x': 2,
+          '_id': 123
+      }])
+                | "Generate ID" >> beam.ParDo(_GenerateObjectIdFn())
+                | "Check" >> beam.Map(lambda x: '_id' in x))
+      assert_that(output, equal_to([True] * 2))
+
+
+class Test_WriteMongoFn(unittest.TestCase):
+  @mock.patch('apache_beam.io.mongodbio._MongoSink')
+  def test_process(self, mock_sink):
+    docs = [{'x': 1}, {'x': 2}, {'x': 3}]
```

Comment:
Consider adding an exhaustive test for dynamic work rebalancing similar to following (not entirely sure if this is possible with mocks).
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/textio_test.py#L386

Reply:
This appears to be a source test? I've added one accordingly to MongoSourceTest

Reply:
Yes. Thanks.